src/catalog/catalog_set.cpp

```diff
@@ -172,162 +236,173 @@ optional_ptr<CatalogEntry> CatalogSet::GetEntryInternal(CatalogTransaction trans
+	renamed_node->timestamp = transaction.transaction_id;
+	renamed_node->deleted = false;
+	renamed_node->set = this;
+	return CreateEntryInternal(transaction, new_name, std::move(renamed_node), read_lock);
+}
+
 bool CatalogSet::AlterEntry(CatalogTransaction transaction, const string &name, AlterInfo &alter_info) {
 	// lock the catalog for writing
-	lock_guard<mutex> write_lock(catalog.GetWriteLock());
+	unique_lock<mutex> write_lock(catalog.GetWriteLock());
 	// lock this catalog set to disallow reading
-	lock_guard<mutex> read_lock(catalog_lock);
+	unique_lock<mutex> read_lock(catalog_lock);
 
-	// first check if the entry exists in the unordered set
-	EntryIndex entry_index;
-	auto entry = GetEntryInternal(transaction, name, &entry_index);
+	// If the entry does not exist, we error
+	auto entry = GetEntryInternal(transaction, name);
 	if (!entry) {
 		return false;
 	}
 	if (!alter_info.allow_internal && entry->internal) {
 		throw CatalogException("Cannot alter entry \"%s\" because it is an internal system entry", entry->name);
 	}
-
-	// create a new entry and replace the currently stored one
-	// set the timestamp to the timestamp of the current transaction
-	// and point it to the updated table node
-	string original_name = entry->name;
 	if (!transaction.context) {
 		throw InternalException("Cannot AlterEntry without client context");
 	}
+
+	// Use the existing entry to create the altered entry
 	auto &context = *transaction.context;
 	auto value = entry->AlterEntry(context, alter_info);
 	if (!value) {
 		// alter failed, but did not result in an error
 		return true;
 	}
 
-	if (value->name != original_name) {
-		auto mapping_value = GetMapping(transaction, value->name);
-		if (mapping_value && !mapping_value->deleted) {
-			auto &original_entry = GetEntryForTransaction(transaction, *mapping_value->index.GetEntry());
-			if (!original_entry.deleted) {
-				entry->UndoAlter(context, alter_info);
-				string rename_err_msg =
-				    "Could not rename \"%s\" to \"%s\": another entry with this name already exists!";
-				throw CatalogException(rename_err_msg, original_name, value->name);
-			}
-		}
-	}
-
-	if (value->name != original_name) {
-		// Do PutMapping and DeleteMapping after dependency check
-		PutMapping(transaction, value->name, entry_index.Copy());
-		DeleteMapping(transaction, original_name);
-	}
-
+	// Mark this entry as being created by this transaction
 	value->timestamp = transaction.transaction_id;
 	value->set = this;
-	auto new_entry = value.get();
-	PutEntry(std::move(entry_index), std::move(value));
 
-	// serialize the AlterInfo into a temporary buffer
-	MemoryStream stream;
-	BinarySerializer serializer(stream);
-	serializer.Begin();
-	serializer.WriteProperty(100, "column_name", alter_info.GetColumnName());
-	serializer.WriteProperty(101, "alter_info", &alter_info);
-	serializer.End();
+	if (!StringUtil::CIEquals(value->name, entry->name)) {
```

Comment:
Can we add a test where we rename only differing in case, e.g.:

```sql
CREATE TABLE MY_TABLE(i int);
ALTER TABLE MY_TABLE RENAME TO my_table;
```